_posts/2018-01-09-dagger2-android-subcomponents.md

```diff
@@ -0,0 +1,205 @@
+            return chain.proceed(request);
+
+        });
+
+        // Logging
+        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
+        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
+        builder.interceptors().add(interceptor);
+
+        Gson gson = new GsonBuilder()
+                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
+                .setDateFormat("yyyy-MM-dd")
+                .create();
+
+        RxJava2CallAdapterFactory rxAdapter = RxJava2CallAdapterFactory
+                .createWithScheduler(Schedulers.io());
+
+        Retrofit retrofit = new Retrofit.Builder()
+                .baseUrl(BuildConfig.API_URL)
+                .client(builder.build())
+                .addConverterFactory(GsonConverterFactory.create(gson))
+                .addCallAdapterFactory(rxAdapter)
+                .build();
+        return retrofit.create(ApiService.class);
+    }
+}
+```
+
+Except for almost standard `Retrofit` initialisation code, there are a few annotations which need special mention 🎉
+##### 1. @Module
+`@Module` tells `Dagger2` that following class is a `Module` and it will provide a bunch of dependancies. We already know what a `Module` is so Nothing fancy here!
+
+##### 2. @Provides
+`@Provides` is an annotation which is for methods in a `Module`. It tells `Dagger2` that following method is responsible for constructing a dependancy and it knows how to its job. `Dagger2` will just invoke this method whenever a dependancy specified with method's return type is required. Note that method name here does not matter, it can be anything just the return type matters.
+
+##### 3. @Singleton
+Last but most interesting annotation in above code is `@Singleton`. This annotation specifies scope of the dependancy being provided. `Singleton` is most wide scope that can be used in `Dagger2`. A `Singleton` dependancy is initialised once **during `Component`'s lifecycle** in `Dagger2`.
+
+That's that, since now we have all the basics and our dependancies ready for injection lets go ahead and create our top level component called `AppComponent` in movie example.
+
+Here is the code for simplified version of `AppComponent`
+```java
+@Singleton
+@Component(modules = {AppModule.class, ApiModule.class})
+public interface AppComponent {
+  void inject(BaseApplication baseApplication);
+}
+```
+Lets have a look at new annotation introduced here
+
+##### @Component
+`@Component` tells dagger that following interface is a `Component` and you need to generate implementation for it. As soon as you write `@Component` and hit build in Android Studio, `Dagger2` generates implementation for the target interface. `Dagger2` adds a prefix `Dagger` in name of implementation class. For an example for our `AppComponent` dagger will generate a class called `DaggerAppComponent`. Also as it apparent from code `@Component` also takes modules which will construct and provide dependancies.
```

Comment:
Should we also mention the details about the inner workings of dagger? how it creates a directed acyclic graph for every component?

Also, it might be a good idea to explain why we have the @singleton annotation on the component as well.